From 57b609f84df9697ce7fbecefd41c9ec9f9c954f3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?J=C3=BCrg=20Billeter?= <j@bitron.ch>
Date: Sat, 23 Apr 2011 16:32:51 +0200
Subject: [PATCH] cryptsetup: order cryptsetup service after udev service

---
 src/cryptsetup-generator.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/src/cryptsetup-generator.c b/src/cryptsetup-generator.c
index 696f44a..57a6999 100644
--- a/src/cryptsetup-generator.c
+++ b/src/cryptsetup-generator.c
@@ -111,7 +111,7 @@ static int create_disk(
                 "Conflicts=umount.target\n"
                 "DefaultDependencies=no\n"
                 "BindTo=%s dev-mapper-%%i.device\n"
-                "After=systemd-readahead-collect.service systemd-readahead-replay.service %s\n"
+                "After=udev.service systemd-readahead-collect.service systemd-readahead-replay.service %s\n"
                 "Before=umount.target\n",
                 d, d);
 
-- 
1.7.4.2

